server.js

```diff
@@ -0,0 +1,247 @@
+      }
+    });
+  }
+
+  // POST
+
+  if (req.method === 'POST' && req.url === '/elements') {
+    if (req.headers.authorization) {
+      const encodedString = req.headers.authorization.slice(6);
+      const base64Buffer = new Buffer(encodedString, 'base64');
+      const decodedString = base64Buffer.toString();
+      let username = decodedString.slice(0, decodedString.indexOf(':'));
+      let password = decodedString.slice(decodedString.indexOf(':') + 1);
+
+      if (authorizedUsers.hasOwnProperty(username) && authorizedUsers[username] === password) {
+        let body = '';
+
+        req.on('data', (chunk) => {
+          body += chunk;
+        });
+
+        req.on('end', (chunk) => {
+          if (chunk) {
+            body += chunk;
+          } else {
+            let parsedBody = qs.parse(body);
+            let pageTemplate = `<!DOCTYPE html>
+        <html lang="en">
+          <head>
+            <meta charset="UTF-8" />
+            <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+            <meta http-equiv="X-UA-Compatible" content="ie=edge" />
+            <title>Document</title>
+            <link rel="stylesheet" href="css/styles.css" />
+          </head>
+          <body>
+          <h1>${parsedBody.elementName}</h1>
+          <h2>${parsedBody.elementSymbol}</h2>
+          <h3>${parsedBody.elementAtomicNumber}</h3>
+          <p>${parsedBody.elementDescription}</p>
+          <p><a href="/">back</a></p>
+          </body>
+        </html>`;
+            let elementName = parsedBody.elementName.toLowerCase();
+            fs.writeFile(`./public/${elementName}.html`, pageTemplate, (err, data) => {
+              if (err) {
+                res.writeHead(500);
+                return res.end('{ "success": false }');
+              } else {
+                updateIndex();
+              }
+              res.writeHead(200, { 'Content-Type': 'application/json' });
+              return res.end('{ "success": true }');
+            });
+          }
+        });
+      } else {
+        res.writeHead(401, { 'WWW-Authenticate': 'Basic realm="SecureArea"' });
+        return res.end(`<html><body>Invalid Authentication Credentials</body></html>`);
+      }
+    } else {
+      res.writeHeader(401, { 'WWW-Authenticate': 'Basic realm="SecureArea"' });
+      return res.end(`<html><body>Not Authorized</body></html>`);
+    }
+  }
+
+  // PUT
+  if (req.method === 'PUT') {
+    if (req.headers.authorization) {
+      const encodedString = req.headers.authorization.slice(6);
+      const base64Buffer = new Buffer(encodedString, 'base64');
+      const decodedString = base64Buffer.toString();
+      let username = decodedString.slice(0, decodedString.indexOf(':'));
+      let password = decodedString.slice(decodedString.indexOf(':') + 1);
+
+      if (authorizedUsers.hasOwnProperty(username) && authorizedUsers[username] === password) {
+        fs.readdir(`./public`, (err, files) => {
```

Comment:
The fs.readdir should be in your res.end. This is readdir is happening the same time as your red.on('data') and your res.end() and it's not guaranteed which will finish first.